### PR TITLE
Cap the Light Attenuation

### PR DESCRIPTION
**Pull request by reduz**
* Given an inverse function is now used, light can reach near-infinity levels.
* This causes problems with auto exposure, volumetric fog, etc.
* Added a new max energy parameter to cap the light to this, used in forward pass as well as volumetric fog.
* GI does not need it because it's capped by design.

___

- *Production edit: Follow-up to https://github.com/godotengine/godot/pull/44941.*

### CODE REVIEW

**Comment by fire** [comment]
Is the cap of 64 supposed to be what is expected.

I did some research:

image

**Comment by clayjohn** [comment]
I'm not sure about this. The ``max(distance, 0.0001)`` part is already designed to clamp to a max brightness, IMO we should be adjusting that rather than adding a new clamp.

**Comment by reduz** [comment]
@clayjohn that is not meant to clamp, its meant to avoid division by zero and its very unfriendly to ask the user to set this value as you don't really even know what the actual intensity will be. It's fine to have both limits.

@fire We are not using physical light units at the moment. By the time we do that limit will appear as such.

File: servers/rendering/renderer_rd/shaders/volumetric_fog.glsl [hunk 1]
```diff
@@ -503,7 +503,7 @@ void main() {
 				float shadow_attenuation = 1.0;
 
 				if (d * spot_lights.data[light_index].inv_radius < 1.0) {
-					float attenuation = get_omni_attenuation(d, spot_lights.data[light_index].inv_radius, spot_lights.data[light_index].attenuation);
+					float attenuation = get_omni_attenuation(d, spot_lights.data[light_index].inv_radius, spot_lights.data[light_index].attenuation, omni_lights.data[light_index].max_energy);
```
Comment (DarkKilauea) [inline servers/rendering/renderer_rd/shaders/volumetric_fog.glsl:506]: Typo, `omni_lights.data[light_index].max_energy` should likely be `spot_lights.data[light_index].max_energy`.

**Comment by s-ilent** [comment]
I feel like this is a bad parameterization.  

How does the user know what the relation between the max_energy and the actual output is? Because of the inverse square law, the energy at the centre will be super high, but the energy at the edges will be comparatively much much smaller. 

>Given an inverse function is now used, light can reach near-infinity levels.

In other engines, the minimum size of point lights is capped at 0.01 (1cm), which means brightest value of an untinted light is 10000. From this code, the minimum cap for attenuation seems to be 0.0001 which (if I'm understanding right) clayjohn says is the same thing. Allowing the user to set the minimum size of lights seems like it would be a better way of parameterising this functionality and, from what I understand, be a simpler calculation as it would only be adjusting that minimum size cap.

https://geom.io/bakery/wiki/index.php?title=Point_Light_Attenuation

I think it would be a good idea to consider something like the "Falloff min size" parameter described on this page. In my experience, it is intuitive to work with, because it deals with the area of the centre of the light, rather than the maximum energy of the attenuation before being tinted (which doesn't have a linear or easy to understand falloff).

**Comment by reduz** [comment]
@s-ilent makes sense, maybe besides changing the limit in the forward pass, I could use a smaller distance for the fog.. let me give it a try.

**Comment by clayjohn** [comment]
@s-ilent The trouble with just doing the size parametrization is that energy is still impacted by the decay (attenuation) parameter. Godot provides an attenuation parameter for artistic control that doesn't appear in the other engines. with attenuation at 2, a min size of 1 cm has a max energy of 10000, but with attenuation at 1, the max energy will just be 100.

**Comment by reduz** [comment]
ah right, I forgot about the attenuation

**Comment by reduz** [comment]
I think there is not much of a merit to this, given @clayjohn is rewritting this code

**Comment by akien-mga** [comment]
Alright, let's close.

**Pull request closed without merging**